modules/components/src/DropDown/MultiSelectDropDown.js

```diff
@@ -0,0 +1,319 @@
+}) => {
+  const [isOpen, setIsOpen] = useState(false);
+  const [selectedItemsMap, setSelectedItemsMap] = useState(new Map());
+  const [defaultColumnsMap, setDefaultColumnsMap] = useState(new Map());
+  const [instanceNum, setInstanceNum] = useState(0);
+  const renderRef = useRef();
+  const buttonRef = useRef();
+  const itemsRef = useRef([]);
+
+  const toggle = (item) => {
+    if (selectedItemsMap.get(item.field)) {
+      // set to null instead of using .delete() because .delete() returns bool
+      setSelectedItemsMap(new Map(selectedItemsMap.set(item.field, null)));
+    } else {
+      setSelectedItemsMap(new Map(selectedItemsMap.set(item.field, item)));
+    }
+
+    if (onChange) {
+      onChange(item);
+    }
+  };
+
+  const selectAll = () => {
+    setSelectedItemsMap(
+      items.reduce((map, item) => {
+        map.set(item.field, item);
+        return map;
+      }, new Map()),
+    );
+
+    if (onMultipleChange) {
+      const toggled = items
+        .filter((item) => !item.show)
+        .reduce((obj, item) => {
+          return { ...obj, [item.field]: true };
+        }, {});
+
+      onMultipleChange(toggled);
+    }
+  };
+
+  const resetToDefaults = () => {
+    setSelectedItemsMap(new Map(defaultColumnsMap));
+
+    if (onMultipleChange) {
+      const toggled = items.filter((item) => {
+        return (
+          (item.show && !defaultColumnsMap.has(item.field)) ||
+          (!item.show && defaultColumnsMap.has(item.field))
+        );
+      });
+
+      const changes = toggled.reduce((obj, item) => {
+        return { ...obj, [item.field]: !item.show };
+      }, {});
+
+      onMultipleChange(changes);
+    }
+  };
+
+  const focusNext = (i) => {
+    itemsRef.current[(i + 1) % items.length].focus();
+  };
+
+  const focusPrev = (i) => {
+    itemsRef.current[(i - 1 + items.length) % items.length].focus();
+  };
+
+  const focusFirst = () => {
+    itemsRef.current[0].focus();
+  };
+
+  const focusLast = () => {
+    itemsRef.current[items.length - 1].focus();
+  };
+
+  const handleKeyPress = (e, i) => {
+    switch (e.key) {
+      case 'ArrowDown':
+        e.preventDefault();
+        focusNext(i);
+        break;
+      case 'ArrowUp':
+        e.preventDefault();
+        focusPrev(i);
+        break;
+      case 'Home':
+        e.preventDefault();
+        focusFirst();
+        break;
+      case 'End':
+        e.preventDefault();
+        focusLast();
+        break;
+      default:
+        break;
+    }
+  };
+
+  const handleEsc = (e) => {
+    if (isOpen && e.key === 'Escape') {
+      setIsOpen(false);
+      buttonRef.current.focus();
+    }
+  };
+
+  const handleBlur = (e) => {
+    const currentTarget = e.currentTarget;
+    // blur happens before click, preventing any click events in children from firing due to rerender from state change
+    // so wait a tick for child component events to fire before changing open state and causing rerender
+    window.setTimeout(() => {
```

Comment:
With event handling, I believe we can take an action without preventing other events from firing. Returning true will allow the event to continue and other handlers to take actions. So we shouldn't need this arbitrary delay.

Reply:
PS. thank you for comments explaining the timeout.